app/models/user.rb

```diff
@@ -379,8 +379,22 @@ def logged?
   end
 
   # user must be assigned all given roles in order to delegate them
-  def can_assign?(roles)
-    can_change_admin_flag? || roles.all? { |r| self.role_ids_was.include?(r) }
+  # or have :escalate_roles permission
+  # but cannot escalate roles for self
+  def can_assign?(role_ids, target_user)
```

Comment:
Disabling login is a way to get to the resources you otherwise would not be able to. But the fact that someone can disable the whole user authentication system should not be a reason why our roles and permissions do not work properly when enabled.

I think preventing an escalation is in this case a necessity. One of the requirements was that the role will be able to assign roles without having them and without self assignment prevented, user having it could modify itself to access almost anything.

You are right that there are cases that need to be fixed but those are existing bugs and I would prefer not to knowingly add more of them. When I tested, I was not able to escalate my roles/permissions as 'Canned admin', so I kept the escalation check for now but I can remove it if you think it should be added together with the fixes for the remaining cases.

Reply:
ok, sounds good and we can keep it in, I'll focus on this while testing too, but I think if system admin can create users and change other users permissions, she can escalate through second account easily